### PR TITLE
Merge master into meson branch - adds HiGHS as submodule

### PR DESCRIPTION
**Pull request by rgommers**

### CODE REVIEW

**Comment by mckib2** [comment]
I have meson installed, building, and reproducing the warning messages.  Have to leave it right now but hope to make more headway tonight

**Comment by rgommers** [comment]
This failed with:
```
FAILED: scipy/optimize/_highs/_highs_wrapper.cpython-39-x86_64-linux-gnu.so.p/.._..__lib_highs_src_io_FilereaderLp.cpp.o 
ccache c++ -Iscipy/optimize/_highs/_highs_wrapper.cpython-39-x86_64-linux-gnu.so.p -Iscipy/optimize/_highs -I../scipy/optimize/_highs -I/opt/hostedtoolcache/Python/3.9.9/x64/lib/python3.9/site-packages/numpy/core/include -I../scipy/optimize/_highs/cython/src -I../scipy/optimize/_highs/src -I../scipy/_lib/highs/extern -I../scipy/_lib/highs/src -I../scipy/_lib/highs/src/io -I../scipy/_lib/highs/src/ipm/ipx/include -I../scipy/_lib/highs/src/lp_data -I../scipy/_lib/highs/src/util -I../scipy/_lib/highs/src/ipm/basiclu/include -I/opt/hostedtoolcache/Python/3.9.9/x64/include/python3.9 -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -Werror -std=c++14 -O2 -g -fPIC -Wno-unused-variable -Wno-unused-but-set-variable '-DCMAKE_BUILD_TYPE="RELEASE"' -DHiGHSRELEASE '-DIPX_ON="ON"' '-DHIGHS_GITHASH="n/a"' '-DHIGHS_COMPILATION_DATE="2021-07-09"' -DHIGHS_VERSION_MAJOR=1 -DHIGHS_VERSION_MINOR=0 -DHIGHS_VERSION_PATCH=0 -DHIGHS_DIR=/home/runner/work/scipy/scipy/scipy/optimize/_highs/../../_lib/highs -UOPENMP -UEXT_PRESOLVE -USCIP_DEV -UHiGHSDEV -UOSI_FOUND -DNPY_NO_DEPRECATED_API=NPY_1_9_API_VERSION -MD -MQ scipy/optimize/_highs/_highs_wrapper.cpython-39-x86_64-linux-gnu.so.p/.._..__lib_highs_src_io_FilereaderLp.cpp.o -MF scipy/optimize/_highs/_highs_wrapper.cpython-39-x86_64-linux-gnu.so.p/.._..__lib_highs_src_io_FilereaderLp.cpp.o.d -o scipy/optimize/_highs/_highs_wrapper.cpython-39-x86_64-linux-gnu.so.p/.._..__lib_highs_src_io_FilereaderLp.cpp.o -c ../scipy/_lib/highs/src/io/FilereaderLp.cpp
../scipy/_lib/highs/src/io/FilereaderLp.cpp: In member function ‘virtual FilereaderRetcode FilereaderLp::readModelFromFile(const HighsOptions&, std::string, HighsModel&)’:
../scipy/_lib/highs/src/io/FilereaderLp.cpp:59:24: error: comparison of integer expressions of different signedness: ‘HighsInt’ {aka ‘int’} and ‘std::vector<std::shared_ptr<Variable> >::size_type’ {aka ‘long unsigned int’} [-Werror=sign-compare]
   59 |     if (num_continuous == m.variables.size()) lp.integrality_.clear();
      |         ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
```
as it should have when using `--werror` in the Linux build.

**Comment by rgommers** [comment]
Non-Meson CI got re-enabled after I installed Azure integration for some unrelated debugging - will get rid of that again.

**Comment by rgommers** [comment]
Hmm, this one is new, and slightly worrying:
```
[1313/1555] Compiling C++ object scipy/optimize/_highs/_highs_wrapper.cp39-win_amd64.pyd.p/.._..__lib_highs_src_lp_data_HighsOptions.cpp.obj
CreateProcess failed. Command attempted:
[1314/1555] Compiling C++ object scipy/optimize/_highs/_highs_wrapper.cp39-win_amd64.pyd.p/.._..__lib_highs_src_mip_HighsConflictPool.cpp.obj
""c++" "-Iscipy\optimize\_highs\_highs_wrapper.cp39-win_amd64.pyd.p" "-Iscipy\optimize\_highs" "-I..\scipy\optimize\_highs" "-IC:\hostedtoolcache\windows\Python\3.9.9\x64\lib\site-packages\numpy\core\include" "-I..\scipy\optimize\_highs\cython\src" "-I..\scipy\optimize\_highs\src" "-I..\scipy\_lib\highs\extern" "-I..\scipy\_lib\highs\src" "-I..\scipy\_lib\highs\src\io" "-I..\scipy\_lib\highs\src\ipm\ipx\include" "-I..\scipy\_lib\highs\src\lp_data" "-I..\scipy\_lib\highs\src\util" "-I..\scipy\_lib\highs\src\ipm\basiclu\include" "-IC:\hostedtoolcache\windows\Python\3.9.9\x64\Include" "-fdiagnostics-color=always" "-D_FILE_OFFSET_BITS=64" "-Wall" "-Winvalid-pch" "-Wnon-virtual-dtor" "-std=c++14" "-O2" "-g" "-DMS_WIN64" "-Wno-unused-variable" "-Wno-unused-but-set-variable" "-DCMAKE_BUILD_TYPE=\"RELEASE\"" "-DHiGHSRELEASE" "-DIPX_ON=\"ON\"" "-DHIGHS_GITHASH=\"n/a\"" "-DHIGHS_COMPILATION_DATE=\"2021-07-09\"" "-DHIGHS_VERSION_MAJOR=1" "-DHIGHS_VERSION_MINOR=0" "-DHIGHS_VERSION_PATCH=0" "-DHIGHS_DIR=D:/a/scipy/scipy/scipy/optimize/_highs/../../_lib/highs" "-UOPENMP" "-UEXT_PRESOLVE" "-USCIP_DEV" "-UHiGHSDEV" "-UOSI_FOUND" "-DNPY_NO_DEPRECATED_API=NPY_1_9_API_VERSION" -MD -MQ scipy/optimize/_highs/_highs_wrapper.cp39-win_amd64.pyd.p/.._..__lib_highs_src_simplex_HSimplexNla.cpp.obj -MF "scipy\optimize\_highs\_highs_wrapper.cp39-win_amd64.pyd.p\.._..__lib_highs_src_simplex_HSimplexNla.cpp.obj.d" -o scipy/optimize/_highs/_highs_wrapper.cp39-win_amd64.pyd.p/.._..__lib_highs_src_simplex_HSimplexNla.cpp.obj "-c" ../scipy/_lib/highs/src/simplex/HSimplexNla.cpp"
ninja: fatal: CreateProcess: The paging file is too small for this operation to complete.
```

**Comment by rgommers** [comment]
tl;dr compiling `highs_wrapper` requires too much memory in the CI job. That may be problematic in other instances as well. Until recently, the `sparsetools` extension took the most memory. We need to measure this and see if it's excessive.

**Comment by mckib2** [comment]
That's surprising - I'm sure there are much larger libraries being compiled in other projects

I could never quite get it to work with distutils - but should certainly be possible with meson - is splitting out the ipx and highs libraries from highs_wrapper. basiclu could trivially be linked as it was a C library, but ipx/highs are C++ libraries and I remember trying very hard but failing to build those separately from highs_wrapper.

Right now I believe there are only two libraries: basiclu and highs_wrapper. For meson I propose splitting out info 4 libraries: basiclu, ipx, highs, highs_wrapper. That should avoid compiling all objects at once and ease the memory burden

**Comment by rgommers** [comment]
It doesn't seem too bad indeed. I tried to measure in two ways:
```
$ /usr/bin/time -v ninja -C build
...
Maximum resident set size (kbytes): 642708
```
And `top` says ~2.5% of total memory (out of 32 GB). So that seems fairly consistent, and we wouldn't expect that to be problematic. https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources says runners have 7 GB RAM.

> Right now I believe there are only two libraries: basiclu and highs_wrapper. For meson I propose splitting out info 4 libraries: basiclu, ipx, highs, highs_wrapper. That should avoid compiling all objects at once and ease the memory burden

That may be a good idea anyway. The build is now kind of heavy:

image

And with `ninja -C build -j 1`:

image

So cumulative build time for the extension is more than 4 minutes, and single-file time 18 seconds:(

**Comment by rgommers** [comment]
Note that I'm testing on Linux though, and the out-of-memory issue is happening on Windows. The Mingw compilers may be worse than GCC on Linux.

**Comment by mckib2** [comment]
> So cumulative build time for the extension is more than 4 minutes, and single-file time 18 seconds:(

That does seem longer than I would have thought.  I just split up libraries in https://github.com/rgommers/scipy/pull/110, if that doesn't significantly improve the results I can try to trim the fat of things we might not need (tests, QP solver, parallelization utilities, etc.)

**Comment by rgommers** [comment]
Thanks. It does seem a little surprising indeed - most of the code seems quite simple, and not templated. There's just a lot of it. The `presolve/HPresolve.cpp` file that's heaviest is ~6300 LoC and includes a lot more.

Maybe worth doing separately though, it shouldn't be a blocker for the merge into SciPy master if all that's failing is a Windows CI job on my `meson` branch.

**Comment by rgommers** [comment]
Hmm, now CI jobs are hanging. It's possible that the out-of-memory probably was due to a temporary hiccup in CI infra.

**Comment by rgommers** [comment]
GitHub is having problems; CI now failing to fetch submodules.

**Comment by rgommers** [comment]
HiGHS files aren't in the top 3 of individual file build times, but if you click through in the trace profiler you see that it contributes 170/1557 targets, and the total build time is high - probably second largest contributor after `scipy.sparse`.

image

So if the number of files that can be removed from the build is significant, that may be worth doing at some point.

**Comment by rgommers** [comment]
Hmm, CI is back to normal, but the Windows job crashed on HiGHS code:
```
......                                                                   [ 40%]
Windows fatal exception: access violation

Current thread 0x00001470 (most recent call first):
  File "D:\a\scipy\scipy\build\Lib\site-packages\scipy\optimize\_linprog_highs.py", line 367 in _linprog_highs
  File "D:\a\scipy\scipy\build\Lib\site-packages\scipy\optimize\_linprog.py", line 618 in linprog
  File "D:\a\scipy\scipy\build\Lib\site-packages\scipy\optimize\tests\test_linprog.py", line 288 in test_choose_solver
  File "C:\hostedtoolcache\windows\Python\3.9.9\x64\lib\site-packages\_pytest\python.py", line 183 in pytest_pyfunc_call
```

**Comment by mckib2** [comment]
> File "D:\a\scipy\scipy\build\Lib\site-packages\scipy\optimize\tests\test_linprog.py", line 288 in test_choose_solver

I'll have to try getting a Windows build working again...  I might have time tomorrow to try it out.  Is there a more simplistic Windows CI job I could look at to try to get things working?

**Comment by rgommers** [comment]
> I'll have to try getting a Windows build working again... I might have time tomorrow to try it out. Is there a more simplistic Windows CI job I could look at to try to get things working?

That'd be great. Not sure what you mean, this one is pretty simple: https://github.com/rgommers/scipy/blob/meson/.github/workflows/windows.yml. It's installing a Mingw toolchain, downloading an openblas binary, pointing `pkg-config` at openblas, and everything should build.

Just to eliminate that the segfault is due to the last commit, let me try and revert that.

**Comment by mckib2** [comment]
> Not sure what you mean, this one is pretty simple:

Got it - I was looking an azure dev ops pipeline before and running into issues. I'll try following the one you pointed out.

Have we seen these segfaults in master or just the meson build? I could believe there's subtleties in linking libraries together correctly on windows - I just don't know them yet!

**Comment by rgommers** [comment]
> Have we seen these segfaults in master or just the meson build? I could believe there's subtleties in linking libraries together correctly on windows - I just don't know them yet!

Just once, on your last commit - I think it's due to the splitting off of more static libraries. That's not urgent (assuming the memory issue yesterday was a flake), so if the revert works we could leave that for later perhaps.

**Comment by rgommers** [comment]
Hmm, the revert is also unhappy on Windows:
```
2021-12-28T10:47:10.7384600Z [1414/1555] Compiling C++ object scipy/optimize/_highs/_highs_wrapper.cp39-win_amd64.pyd.p/.._..__lib_highs_src_simplex_HSimplexDebug.cpp.obj
2021-12-28T10:47:12.4636844Z FAILED: scipy/optimize/_highs/_highs_wrapper.cp39-win_amd64.pyd.p/.._..__lib_highs_src_simplex_HSimplexDebug.cpp.obj 
2021-12-28T10:47:12.4645323Z "c++" "-Iscipy\optimize\_highs\_highs_wrapper.cp39-win_amd64.pyd.p" "-Iscipy\optimize\_highs" "-I..\scipy\optimize\_highs" "-IC:\hostedtoolcache\windows\Python\3.9.9\x64\lib\site-packages\numpy\core\include" "-I..\scipy\optimize\_highs\cython\src" "-I..\scipy\optimize\_highs\src" "-I..\scipy\_lib\highs\extern" "-I..\scipy\_lib\highs\src" "-I..\scipy\_lib\highs\src\io" "-I..\scipy\_lib\highs\src\ipm\ipx\include" "-I..\scipy\_lib\highs\src\lp_data" "-I..\scipy\_lib\highs\src\util" "-I..\scipy\_lib\highs\src\ipm\basiclu\include" "-IC:\hostedtoolcache\windows\Python\3.9.9\x64\Include" "-fdiagnostics-color=always" "-D_FILE_OFFSET_BITS=64" "-Wall" "-Winvalid-pch" "-Wnon-virtual-dtor" "-std=c++14" "-O2" "-g" "-DMS_WIN64" "-Wno-unused-variable" "-Wno-sign-compare" "-Wno-switch" "-Wno-format-truncation" "-Wno-non-virtual-dtor" "-Wno-class-memaccess" "-Wno-unused-but-set-variable" "-DCMAKE_BUILD_TYPE=\"RELEASE\"" "-DHiGHSRELEASE" "-DIPX_ON=\"ON\"" "-DHIGHS_GITHASH=\"n/a\"" "-DHIGHS_COMPILATION_DATE=\"2021-07-09\"" "-DHIGHS_VERSION_MAJOR=1" "-DHIGHS_VERSION_MINOR=0" "-DHIGHS_VERSION_PATCH=0" "-DHIGHS_DIR=D:/a/scipy/scipy/scipy/optimize/_highs/../../_lib/highs" "-UOPENMP" "-UEXT_PRESOLVE" "-USCIP_DEV" "-UHiGHSDEV" "-UOSI_FOUND" "-DNPY_NO_DEPRECATED_API=NPY_1_9_API_VERSION" -MD -MQ scipy/optimize/_highs/_highs_wrapper.cp39-win_amd64.pyd.p/.._..__lib_highs_src_simplex_HSimplexDebug.cpp.obj -MF "scipy\optimize\_highs\_highs_wrapper.cp39-win_amd64.pyd.p\.._..__lib_highs_src_simplex_HSimplexDebug.cpp.obj.d" -o scipy/optimize/_highs/_highs_wrapper.cp39-win_amd64.pyd.p/.._..__lib_highs_src_simplex_HSimplexDebug.cpp.obj "-c" ../scipy/_lib/highs/src/simplex/HSimplexDebug.cpp
2021-12-28T10:47:13.6837926Z c++: fatal error: cannot execute 'c:/rtools40/ucrt64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/bin/as.exe': CreateProcess: No such file or directory
```

**Comment by rgommers** [comment]
Putting back the `static_library` commit. It's so flaky that I think it'd indeed be good to debug locally on Windows if you can @mckib2.

**Comment by rgommers** [comment]
Same `Windows fatal exception: access violation` still - it's reproducible and not due to CI infra.

**Comment by rgommers** [comment]
Hmm, now the memory error on the Windows job is elsewhere:
```
[1311/1557] Compiling Cython source D:/a/scipy/scipy/scipy/ndimage/src/_cytest.pyx
CreateProcess failed. Command attempted:
""cc" "-Iscipy\ndimage\_cytest.cp39-win_amd64.pyd.p" "-Iscipy\ndimage" "-I..\scipy\ndimage" "-IC:\hostedtoolcache\windows\Python\3.9.9\x64\lib\site-packages\numpy\core\include" "-IC:\hostedtoolcache\windows\Python\3.9.9\x64\Include" "-fdiagnostics-color=always" "-D_FILE_OFFSET_BITS=64" "-w" "-std=c99" "-O2" "-g" "-Wno-unused-but-set-variable" "-mlong-double-64" "-DMS_WIN64" "-Wno-unused-function" "-Wno-conversion" "-Wno-misleading-indentation" "-Wno-incompatible-pointer-types" -MD -MQ scipy/ndimage/_cytest.cp39-win_amd64.pyd.p/meson-generated_scipy_ndimage_src__cytest.pyx.c.obj -MF "scipy\ndimage\_cytest.cp39-win_amd64.pyd.p\meson-generated_scipy_ndimage_src__cytest.pyx.c.obj.d" -o scipy/ndimage/_cytest.cp39-win_amd64.pyd.p/meson-generated_scipy_ndimage_src__cytest.pyx.c.obj "-c" scipy/ndimage/_cytest.cp39-win_amd64.pyd.p/scipy/ndimage/src/_cytest.pyx.c"
ninja: fatal: CreateProcess: The paging file is too small for this operation to complete.
```
Current guess: we're likely using too much memory due to the `-j 128`, which is a workaround for another issue.

**Comment by rgommers** [comment]
@mckib2 what do you think about merging this PR in its current state, and then have two follow-ups:
- you can investigate the crash on Windows in the HiGHS tests
- I can work on getting rid of the workaround for the generated Cython code.

My PR to merge Meson support into SciPy master only has a Linux CI job, which will work fine. So merging this PR and syncing it there will enable us to merge that PR finally.

**Comment by mckib2** [comment]
> @mckib2 what do you think about merging this PR in its current state, and then have two follow-ups:

Sounds good to me -- make sure to poke me if you don't hear anything in a little bit, still learning how to juggle time between this and baby!

**Comment by rgommers** [comment]
Great, thanks Nicholas. Juggling time is hard - but baby always first I guess:)
